lib/internal/main/watch_mode.js

```diff
@@ -103,27 +103,25 @@ async function stop() {
   clearGraceReport();
 }
 
+let restarting = false;
 async function restart() {
+  if (restarting) return;
+  restarting = true;
   if (!kPreserveOutput) process.stdout.write(clear);
   process.stdout.write(`${green}Restarting ${kCommandStr}${white}\n`);
   await stop();
   start();
+  restarting = false;
 }
 
-(async () => {
-  emitExperimentalWarning('Watch mode');
-
-  try {
-    start();
-
-    // eslint-disable-next-line no-unused-vars
-    for await (const _ of on(watcher, 'changed')) {
-      await restart();
-    }
-  } catch (error) {
+emitExperimentalWarning('Watch mode');
+start();
+watcher
+  .on('changed', restart)
```

Comment:
note that there is a minor change here, as previously, you'd await `restart` but now it'll throw an unhandled exception through the event emitter if it throws (and won't do watcher.off('changed')). Not sure how much it matters.